creation/error-prone-actor-dependencies.md

```diff
@@ -0,0 +1,128 @@
+
+```pony
+use "files"
+
+actor Writer
+  var _file: (File | None) = None
+
+  new create(env: Env, file_name: String) =>
+    try
+      let auth = env.root as AmbientAuth
+      _file = File(FilePath(auth, file_name))
+    end
+
+  be record(it: String) =>
+    _file.write(it)
+```
+
+We're now left with one more compiler error to address.
+
+```
+x.pony:13:9: couldn't find write in None val
+    _file.write(it)
+         ^
+```
+
+One more change address that `file` could be uninitialized:
+
+```pony
+use "files"
+
+actor Writer
+  var _file: (File | None) = None
+
+  new create(env: Env, file_name: String) =>
+    try
+      let auth = env.root as AmbientAuth
+      _file = File(FilePath(auth, file_name))
+    end
+
+  be record(it: String) =>
+    match _file
+    | let f: File => f.write(it)
+    end
+```
+
+In the final above example, in our `record` behavior, we match on `file` and only attempt to write if it is of type `File`. Awesome. We have working code. Except, ugh. There are actually several problems lurking.
+
+One is programmer ergonomics. If you are using `file` a lot in this actor, you are going to be constantly matching to make sure you are handling `None` correctly.
+
+We are also silently eating failures. If this actor can't start up properly then we might not want to continue running. We aren't going to know that of course. As far as a caller is concerned, we've successfully initialized and we are writing data to the file. What might actually be happening is that every call to `record` results in absolutely nothing. Tracking that down could turn out to be a nightmare. 
+
+Lastly, even if we wanted to communicate failure back, this is an actor. Everything is asynchronous and there's no straightforward way to say something like:
+
+```pony
+if (my_actor.is_initialized()) then
+  my_actor.record(it)
+else
+  error
+end
+```
+
+And even if there was, we want to fail on initilization, not lazily at some unknown time in the future.
+
+## Solution
+
+All of the problems that we enumerated above come from attempting to create objects whose creation can fail in the constructor of our actor. Rather than delay errors until we are in our actor's constructor, a much better approach, is supply our dependencies fully initialized.
+
+In our previous case, we were relying on two things:
+
+* Ambient Authority
+* Existence of our File
+
+If we initialize these outside of our actor then we can easily report construction erorrs and avoid messing with `None` as a possibility inside our `Writer` actor:
+
+```pony
+use "files"
+
+actor Main
+  new create(env: Env) =>
+
+    try
+      let auth = env.root as AmbientAuth
+      let file = recover File(FilePath(auth, "free-candy.txt")) end
+      let writer = Writer(consume file)
+    else
+      env.err.print("Couldn't initialize dependencies")
+    end
+
+actor Writer
+  let _file: File iso
+
```

Comment:
I would make this a `File` instead of a `File iso`, to show that just because you initialise with something isolated, it doesn't have to stay isolated.